### PR TITLE
enhancement/pca estimate of no of signifcant components

### PR DESCRIPTION
**Pull request by pquinn-dls**
I've added a method to estimate the no of signficant components in PCA - enhancement on the current threshold method.
This performs a simple measurements of where the elbow position is and stores it. 
This calculation is performed in the decomposition method and when using plot_explained_variance_ratio  setting vline=True will plot the 
vertical line for this estimated position. 

screeplot3

A property no_significant_components has also been added to LearningResults. Storing the value is to help with automated scripts and later cluster analysis commits.
Small change also added to user guide

### CODE REVIEW

**Comment by jat255** [comment]
@pquinn-dls thanks for the contribution! At first glance, this looks pretty good, and very useful! Is there a reference that you could include in the documentation to using this method? It looks like a fairly obvious method once I read the comments in the code, but I hadn't seen it used before. 

A couple of requests:
- [x] Can you please add a few test cases and ensure that they pass? You can look in the `test` folder for examples using the `pytest` framework. I would create a few dummy signals with a known number of significant components, and then ensure that the method returns the expected result.
- [x] If it's not too much trouble, it might be nice to have the "triangle" visualized at least in the documentation to make it clear how the point is determined. Here's a horrible quick example of what I mean using a screenshot tool:
   - image
- [x] It seems to me that in your example, the 4th component is the furthest from the "triangle line", but the `vline` appears to be drawn through the 5th component. Is this an error, or am I just not understanding how the method works correctly?

**Comment by pquinn-dls** [comment]
Hi,

Yes you're correct the elbow is the 4th but the no of significant pca components was set to elbow + 1
I realised that I was plotting the no of components - not the index so I've adjusted that. 

I've added a reference, some additional docs and an example image of the method.

Ville Satopää, Jeannie Albrecht, David Irwin, Barath Raghavan. Finding a "Kneedle" in a Haystack: Detecting Knee Points in System Behavior. 31st International Conference on Distributed Computing Systems Workshops, pp. 166-171, Minneapolis, Minnesota, USA, June 2011.

I've also added a test but I'm not sure if I've done this correctly - need to read a bit more on this!

File: doc/user_guide/mva.rst
Comment (jat255) [inline doc/user_guide/mva.rst]: `no` --> `number`

File: doc/user_guide/mva.rst
Comment (jat255) [inline doc/user_guide/mva.rst]: "vline" --> \`\`vline\`\`

Comment (jat255) [inline doc/user_guide/mva.rst]: "In this case the elbow" --> "In this case, the elbow"

File: doc/user_guide/mva.rst
Comment (jat255) [inline doc/user_guide/mva.rst]: "no" --> "number"

**Review by jat255 — CHANGES_REQUESTED** [review]

File: doc/user_guide/mva.rst
Comment (pquinn-dls) [reply doc/user_guide/mva.rst]: changed

File: doc/user_guide/mva.rst
Comment (pquinn-dls) [reply doc/user_guide/mva.rst]: fixed

File: doc/user_guide/mva.rst
Comment (pquinn-dls) [reply doc/user_guide/mva.rst]: fixed

File: doc/user_guide/mva.rst
Comment (jat255) [inline doc/user_guide/mva.rst]: Sorry, can you change this to use backticks (\`) instead of single quotes (like \`\`threshold\`\` above it)? That will cause the value to be formatted in monospace in the generated documentation.

**Review by jat255 — CHANGES_REQUESTED** [review]

**Comment by jat255** [comment]
Also, there were some merges into `RELEASE_next_minor`  yesterday. Can you please rebase your branch onto an updated copy of `RELEASE_next_minor`?

**Comment by pquinn-dls** [comment]
Rebased to RELEASE_next_minor and fixed the vline documentation.

**Comment by jat255** [comment]
I made some suggested changes in https://github.com/pquinn-dls/hyperspy/pull/2, and it looks like @ericpre did so as well in https://github.com/pquinn-dls/hyperspy/pull/1. Can you please review these and incorporate them onto your branch? After that I think this PR will be good to go.

File: hyperspy/learn/mva.py
Comment (jat255) [inline hyperspy/learn/mva.py:1283]: I'll add here that it would be better to do this with a vector calculation (https://stackoverflow.com/a/46470401/1435788), rather than looping through the array, but since it's such a small one-dimensional array, there'e probably no harm in leaving it as a loop.

File: doc/user_guide/mva.rst
Comment (tjof2) [inline doc/user_guide/mva.rst:120]: I think more details are needed here rather than just referring back to the paper:

> In this case, the elbow or knee is found in the variance plot by estimating the distance from each point in the variance plot to a line joining the first and last points of the plot.

We should be clear about e.g. choosing between multiple candidates.

Comment (pquinn-dls) [reply doc/user_guide/mva.rst:120]: Updated the description and commented on multiple candidates.

File: hyperspy/learn/mva.py
Comment (pquinn-dls) [reply hyperspy/learn/mva.py:1283]: If it's ok I'll leave it as is for now as I think it's negligible computationally compared to the decomposition process in the majority of cases.

Comment (pquinn-dls) [reply hyperspy/learn/mva.py:1283]: I've merged and edited the changes and comments from @jat255  and @ericpre

**Comment by jat255** [comment]
This looks pretty good to me, but will have to wait for @ericpre or @francisco-dlp to merge, since they have write access.

**Comment by pquinn-dls** [comment]
Great, thanks for your help

**Review by ericpre — COMMENTED** [review]
It looks good to me too. I have one comment: rename `num_significant_components` to `number_significant_components` to follow [PEP8](https://www.python.org/dev/peps/pep-0008/)? 
Even if this can be regarded as a matter of taste, we usually use full words as naming conventions and using `number` instead of `num` doesn't make it significantly longer. I am so used to it now that I was mislead by the original `no_significant_components`.

**Comment by ericpre** [comment]
The failures are related to the releases of upstream libraries (numpy and mrcz).

**Comment by ericpre** [comment]
Thanks @pquinn-dls.

**Pull request merged**
